two phases flow codes/Dambreak_LevelSet_script.py

```diff
@@ -0,0 +1,239 @@
+nn=0
+
+res=50 #resolution
+scale=1
+mesh = RectangleMesh(Point(0,0), Point(scale,scale),res,res,"crossed")
+plot(mesh)
+n = FacetNormal(mesh)
+
+i, j, k, l = indices(4)
+delta = Identity(2)
+
+scalar = FiniteElement('P', mesh.ufl_cell(), 1) #pressure level set function
+vector = VectorElement('P', mesh.ufl_cell(), 2) #velocity
+vector1 = VectorElement('P', mesh.ufl_cell(), 1) #interface normal
+mixed_element = MixedElement([scalar, vector, vector])
+ScalarSpace = FunctionSpace(mesh, scalar) #pressure
+ScalarSpace1 = FunctionSpace(mesh, scalar) #level set function
+VectorSpace = FunctionSpace(mesh, vector) #velocity
+VectorSpace1 = FunctionSpace(mesh, vector1) #interface normal
+
+## Define boundary condition
+
+wallsx = 'on_boundary && (near(x[0], 1, 1e-14) || near(x[0], 0, 1e-14) )' #x=1 and x=0
+wallsy = 'on_boundary && (near(x[1], 1, 1e-14) || near(x[1], 0, 1e-14) )' #y=1 and y=0
+bc = []
+bc2 = []
+# free slip boundary condition dot(u,n)=0
+bc2.append(DirichletBC(VectorSpace.sub(0), Constant(0), wallsx)) 
+bc2.append(DirichletBC(VectorSpace.sub(1), Constant(0), wallsy))
+
+
+v = TestFunction(VectorSpace)
+q = TestFunction(ScalarSpace)
+u_t = Function(VectorSpace) #tentative velocity
+u_c = Function(VectorSpace) #tentative velocity after correction
+u_c0 = Function(VectorSpace) #tentative velocity after correction
+p = Function(ScalarSpace)
+p0 = Function(ScalarSpace)
+
+del_phi_c = TestFunction(ScalarSpace1) #level set function
+phi_c = Function(ScalarSpace1) #level set function
+phi_cr = Function(ScalarSpace1) #level set function after reinitialization
+phi_c0 = Function(ScalarSpace1)
+phi_cr0 = Function(ScalarSpace1) #level set function after reinitialization
+
+del_n_phi = TestFunction(VectorSpace1) 
+n_phi = Function(VectorSpace1) #normal of interface calculate from level set function
+
+V = FunctionSpace(mesh, 'P', 1)
+
+
+## Define initial conditions
+epsilon=0.5*pow(scale/res,0.9)#related to the size of element
+interface_x=0.3*scale
+interface_y=0.5*scale
+initial_phi_c = CharacterLevelSetFunction(param1=interface_x, param2=interface_y, param3=epsilon, element=V.ufl_element(),degree=1)#initial level set function
+initial_v = Expression(('0.', '0.'), degree=1)
+initial_p = Expression('(x[0] < x0 && x[1] < y0) ? (y0-x[1])*1.0 : 0.0',x0=interface_x,y0=interface_y, degree=1) #initial the pressure field of water 
+
+#%%
+## initialization
+
+# project(initial_phi_c, ScalarSpace1)
+# project(initial_v, VectorSpace)
+# project(initial_p, ScalarSpace)
+phi_c.assign(project(initial_phi_c, ScalarSpace1))
+phi_c0.assign(project(initial_phi_c, ScalarSpace1))
+u_c0.assign(project(initial_v, VectorSpace))
+p0.assign(project(initial_p*initial_phi_c, ScalarSpace)) 
+#mutiply the inital level-set value can make the initial calculation step more stable, but it has no effect on the result
+# can also use p0.assign(project(initial_p, ScalarSpace)) 
+
+
+# set parameters
+Re=2000.
+Fr=0.7
+rho0=1.
+nu0=1.
+rho1=0.001
+nu1=1.
```

Comment:
That is quite a big density ratio...

Reply:
Yes, I use the density of water as a standard for dimensionless, and then 
```python
rho0=1. #water
rho1=0.001 #air
```
But for the viscosity, it will significantly affect the convergence, so the water and air are set to be the same.
I don't know if this will be a problem

Reply:
In reality, the viscosities of water and air are not equal [see here](https://en.wikipedia.org/wiki/List_of_viscosities). I can imagine that it does not converge because the viscosities differ by several orders of magnitude. Nevertheless, you should put a not here why the viscosities are equal and explore the convergence problem.